pkg/restapi/vc/operation/operations.go

```diff
@@ -828,39 +830,145 @@ func (o *Operation) issueCredentialHandler(rw http.ResponseWriter, req *http.Req
-		o.writeErrorResponse(rw, http.StatusInternalServerError, err.Error())
+		o.writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(invalidRequestErrMsg+": %s", err.Error()))
+
+		return
+	}
+
+	// create the verifiable credential
+	credential, err := buildCredential(&composeCredReq)
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf("failed to build credential:"+
+			" %s", err.Error()))
 
 		return
 	}
 
+	// sign the credential
+	// TODO https://github.com/trustbloc/edge-service/issues/142 sign using the proofFormat options in the
+	//  request (for now, using issuer val from the request)
+	signedVC, err := o.signCredential(credential, composeCredReq.Issuer)
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf("failed to sign credential:"+
+			" %s", err.Error()))
+
+		return
+	}
+
+	// response
+	rw.WriteHeader(http.StatusOK)
+	o.writeResponse(rw, signedVC)
+}
+
+func (o *Operation) signCredential(credential *verifiable.Credential, didID string) (*verifiable.Credential, error) {
+	// Resolve DID and get the public keyID
+	didDoc, err := o.vdri.Resolve(didID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve DID : %s", err.Error())
+	}
+
+	publicKeyID, err := getPublicKeyID(didDoc)
+	if err != nil {
+		return nil, err
+	}
+
 	// sign the credential
 	signedVC, err := o.crypto.SignCredential(
 		&vcprofile.DataProfile{
 			Creator: publicKeyID,
-			// TODO passed as options to request or set by environment variables ?
+			// TODO https://github.com/trustbloc/edge-service/issues/125 passed as options to request or
+			//  set by environment variables ?
 			SignatureType:           "Ed25519Signature2018",
 			SignatureRepresentation: verifiable.SignatureJWS,
 		},
-		validatedCred,
+		credential,
 	)
 	if err != nil {
-		o.writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf("failed to sign credential: %s", err.Error()))
+		return nil, fmt.Errorf("failed to sign credential: %s", err.Error())
+	}
 
-		return
+	return signedVC, nil
+}
+
+func buildCredential(composeCredReq *ComposeCredentialRequest) (*verifiable.Credential, error) {
+	// create the verifiable credential
+	credential := &verifiable.Credential{}
+
+	// set credential data
+	credential.Context = []string{"https://www.w3.org/2018/credentials/v1"}
+	credential.Issued = composeCredReq.IssuanceDate
+	credential.Expired = composeCredReq.ExpirationDate
+	credential.Evidence = &composeCredReq.Evidence
+
+	// set default type, if request doesn't contain the type
+	credential.Types = []string{"VerifiableCredential"}
+	if len(composeCredReq.Types) != 0 {
+		credential.Types = composeCredReq.Types
 	}
 
-	rw.WriteHeader(http.StatusOK)
-	o.writeResponse(rw, signedVC)
+	// set subject
+	if composeCredReq.Claims == nil {
+		composeCredReq.Claims = make(map[string]interface{})
+	}
+
+	composeCredReq.Claims["subject"] = composeCredReq.Subject
+	credential.Subject = composeCredReq.Claims
```

Comment:
These two lines are confusing. Why are we doing this?

Reply:
Combining `subject` and `claims` into `credentialSubject`.